### PR TITLE
config: add validation for configuration 

### PR DESCRIPTION
**Pull request by jackysp**
As title says. PTAL @coocood @shenli

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -25,6 +25,58 @@ import (
 	tracing "github.com/uber/jaeger-client-go/config"
 )
 
+// Config number limitations
+const (
```
Comment (shenli) [inline config/config.go]: We should document them in config.yaml.template and docs repos.

Comment (jackysp) [reply config/config.go]: Let's check these if these const is suitable or not, then I'll change the template and docs.

File: config/config.go [hunk 2]
```diff
@@ -25,6 +25,58 @@ import (
...
+	MinProcs                 = 0
+	MaxProcs                 = 1000
+	MinRetryLimit            = 0
+	MaxRetryLimit            = 100
+	MinJoinConcurrency       = 1
+	MaxJoinConcurrency       = 100
+	MinStmtCountLimit        = 1000
+	MaxStmtCountLimit        = 1 << 20
+	MinPlanCacheCap          = 1
+	MaxPlanCacheCap          = 10000
```
Comment (zz-jason) [inline config/config.go]: How about 1<<20 ? It can be used as long as our users have large enough memory.

File: config/config.go [hunk 3]
```diff
@@ -25,6 +25,58 @@ import (
...
+	MinPlanCacheShards       = 1
+	MaxPlanCacheShards       = 1000
```
Comment (zz-jason) [inline config/config.go]: 1 << 10 is more efficient

File: config/config.go [hunk 4]
```diff
@@ -25,6 +25,58 @@ import (
...
+	MinPrePlanCacheCap       = 1
+	MaxPrePlanCacheCap       = 1000
```
Comment (zz-jason) [inline config/config.go]: ditto, how about 1<<20 ?

File: config/config.go [hunk 5]
```diff
@@ -234,8 +286,8 @@ var defaultConf = Config{
 		StmtCountLimit:  5000,
 	},
 	XProtocol: XProtocol{
```
Comment (coocood) [inline config/config.go]: We can simply ignore this field.

File: config/config.go [hunk 6]
```diff
@@ -25,6 +25,58 @@ import (
...
+)
+
+// Valid config maps
+var (
+	ValidStorage = map[string]bool{
+		"mocktikv": true,
+		"tikv":     true,
+	}
+	// TODO: Remove it when X Server is available.
+	ValidXServer = map[bool]bool{
```
Comment (coocood) [inline config/config.go]: Why not just use a bool value `false`?

File: config/config.go (hunk 2 above)
Comment (jackysp) [reply config/config.go]: For a global level cache capacity, I think 10000 is large enough.

File: config/config.go (hunk 4 above)
Comment (jackysp) [reply config/config.go]: 1 << 20 is so large for a session.

File: config/config.go (hunk 6 above)
Comment (jackysp) [reply config/config.go]: In my opinion, there are two kinds of limitations for settings. One is the range like, and the other is the list like. I prefer to use a uniform style for the two styles.

**Comment by coocood** [comment]
@jackysp 
I think for the configuration limit, we don't need to make it too strict, just prevent insanity numbers will do.

**Comment by coocood** [comment]
We can have just a few max numbers such as `1 << 10`, `1 << 16`, `1 << 20`, `1 << 32`.

File: tidb-server/main.go [hunk 7]
```diff
@@ -323,6 +323,93 @@ func validateConfig() {
 		log.Error("TiDB run with skip-grant-table need root privilege.")
 		os.Exit(-1)
 	}
+	if cfg.Port < config.MinPortNumber || cfg.Port > config.MaxPortNumber {
+		log.Error("Invalid server port number.")
```
Comment (shenli) [inline tidb-server/main.go]: We should also output the MinPortNumber/MaxPortNumber and cfg.Port in the error message.

File: tidb-server/main.go [hunk 8]
```diff
@@ -323,6 +323,93 @@ func validateConfig() {
...
+		os.Exit(-1)
+	}
+	if _, ok := config.ValidStorage[cfg.Store]; !ok {
+		log.Error("Invalid storage name.")
```
Comment (winkyao) [inline tidb-server/main.go]: ditto, add more useful info in error.

**Comment by jackysp** [comment]
PTAL @coocood @winkyao @shenli @zz-jason

File: config/config.go [hunk 9]
```diff
@@ -25,18 +25,31 @@ import (
 	tracing "github.com/uber/jaeger-client-go/config"
 )
 
+// Config number limitations
+const (
+	MaxLogFileSize = 4096 // MB
+)
+
+// Valid config maps
```
Comment (zimulala) [inline config/config.go]: Add `.` at the end of a line.
Rest LGTM.

File: tidb-server/main.go [hunk 10]
```diff
@@ -323,12 +327,30 @@ func validateConfig() {
 		log.Error("TiDB run with skip-grant-table need root privilege.")
 		os.Exit(-1)
 	}
+	if _, ok := config.ValidStorage[cfg.Store]; !ok {
+		nameList := ""
```
Comment (tiancaiamao) [inline tidb-server/main.go]: There is a `strings.Join` in Go standard library.

See https://godoc.org/strings#Join

**Comment by jackysp** [comment]
PTAL @tiancaiamao

File: tidb-server/main.go [hunk 11]
```diff
@@ -51,6 +51,7 @@ import (
 	"github.com/prometheus/client_golang/prometheus/push"
 	log "github.com/sirupsen/logrus"
 	"google.golang.org/grpc"
+	"strings"
```
Comment (tiancaiamao) [inline tidb-server/main.go]: This line should be moved to the block together with other standard library

File: tidb-server/main.go [hunk 12]
```diff
@@ -323,12 +328,30 @@ func validateConfig() {
 		log.Error("TiDB run with skip-grant-table need root privilege.")
 		os.Exit(-1)
 	}
+	if _, ok := config.ValidStorage[cfg.Store]; !ok {
+		var nameList []string
```
Comment (tiancaiamao) [inline tidb-server/main.go]: `nameList := make([]string, len(config.ValidStorage, 0)`

**Comment by jackysp** [comment]
PTAL @tiancaiamao

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**